### PR TITLE
Track delta metrics between scrapes

### LINKED ISSUE #116: Handle DELTA metrics better
Right now, there are several issues stemming from stack driver's DELTA metrics, which just record the change in a metric. This leads to inconsistent data in prometheus, since the value reported by the exporter is a gauge, which is only set to the value of the change. Ideally, we'd update a gauge with the value of the delta change to reduce the impedance mismatch between stackdriver and prometheus. This is most noticeable with infrequently updated metrics, which produce a single gauge data point in prometheus once in a while, with huge gaps between them.

It looks like DELTA metrics used to be handled as counters, but that was changed with #15 since DELTAs can go down. There probably just needs to be some different handling of how we set the gauge. If it's a GAUGE type from stackdriver, we set the value of the prometheus gauge to the value from stackdriver. If it's a DELTA type from stackdriver, we add the value of the stackdriver delta to the value of the prometheus gauge.

### PR DESCRIPTION
**Pull request by kgeckhart**
This PR introduces support for aggregating DELTA metrics in-memory. This implementation is somewhat based on https://github.com/prometheus-community/stackdriver_exporter/pull/167 but with extra functionality which is intended to solve the problem identified here, https://github.com/prometheus-community/stackdriver_exporter/pull/167#issuecomment-1217127124.

I chose to inject interfaces for the delta and distribution stores at the Collector level to try to keep library friendliness intact, https://github.com/prometheus-community/stackdriver_exporter/pull/157

I did as much as I could to explain the solution and the tradeoffs with aggregating deltas in this manner in the README.md. The bulk of the changes are in `monitoring_metrics.go` where I tried to obfuscate as much of the delta logic as possible.

Should hopefully resolve #116 and #25

### CODE REVIEW

**Comment by SuperQ** [comment]
Thanks, I'll take a look at this soon.

To start, this needs a DCO sign-off. You can use `git commit -s --amend` to add it.

**Comment by ahjmorton** [comment]
Hello @SuperQ / @kgeckhart, this looks like it solves a really important issue with the exporter where DELTA metrics seem pretty much unusable in certain cases. 

Can I ask what is stopping it from being merged and if there is anything I can do to speed that up? This issue is making monitoring of some of our GCP dependencies a lot harder as the data is split into 2 places.

**Comment by kgeckhart** [comment]
Hi @ahjmorton, I'll defer to @SuperQ for the review + merge side of things. I can say we have been running this version in production for a little over three weeks. Our largest exporter is tracking ~500k metrics without any noticeable issues.

File: README.md
Comment (SuperQ) [inline README.md:143]: Just to note, we have an OpenMetrics solution for the starting-at-zero problem. But we need Prometheus support for OpenMetrics created timestamps.

We need this for a bunch of reasons, so I've added it to the Promethus dev summit agenda so we can talk about it again.

File: README.md
Comment (SuperQ) [inline README.md:160]: I wonder if we should follow the Prometheus default and have the staleness TTL be 5 minutes.

File: collectors/delta_counter.go
Comment (SuperQ) [inline collectors/delta_counter.go:1]:
```suggestion
// Copyright 2022 The Prometheus Authors
```

File: collectors/delta_distribution.go
Comment (SuperQ) [inline collectors/delta_distribution.go:1]:
```suggestion
// Copyright 2022 The Prometheus Authors
```

**Review by SuperQ — CHANGES_REQUESTED** [review]
Minor nits, but otherwise LGTM.

**Comment by SuperQ** [comment]
CC @igorwwwwwwwwwwwwwwwwwwww, this may interest you.

**Comment by igorwwwwwwwwwwwwwwwwwwww** [comment]
cc @t4cc0re

File: README.md
Comment (kgeckhart) [reply README.md:160]: I ended up going for a higher default because there are a reasonable amount of metrics with a note of `After sampling, data is not visible for up to x seconds` where x > 5 minutes. 30 minutes might be a bit overzealous/arbitrary so I don't mind dropping it down further. I will say so far it hasn't seemed to cause a large increase in memory usage in our environment.

Comment (SuperQ) [reply README.md:160]: Ok, makes sense. I didn't know Stackdriver would provide such low granularity for some things. The 30m default is fine.

**Review by SuperQ — APPROVED** [review]
LGTM

**Pull request merged**
